### PR TITLE
[LV] Stability fix for outerloop vectorization

### PR DESCRIPTION
**Pull request by nikolaypanchenko**
HCFG builder doesn't correctly handle cases when non-outermost loop is requested to be vectorized

[Original] Differential Revision: https://reviews.llvm.org/D150700

### CODE REVIEW

File: llvm/test/Transforms/LoopVectorize/outer_loop_hcfg_construction.ll [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+;     #pragma clang loop vectorize_width(4)
+;     for (int i = 0; i < n; ++i) {
+;         for (int j = 0; j < n; ++j) {
+;             a[i][j] = 2 + k;
+;         }
+;     }
+;   }
+; }
+
+define void @test(i64 %n, ptr %a) {
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/outer_loop_hcfg_construction.ll:16]: Would it be possible to use more descriptive names for the test and possible add a brief description of what the test checks? (as mentioned in https://reviews.llvm.org/D150700)

File: llvm/test/Transforms/LoopVectorize/outer_loop_hcfg_construction.ll [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+; CHECK-NEXT:    ret void
+;
+entry:
+  %cmp34 = icmp sgt i64 %n, 0
+  br i1 %cmp34, label %for.body.us.preheader, label %for.cond.cleanup
+
+for.body.us.preheader:
+  br label %for.body.us
+
+for.body.us:
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/outer_loop_hcfg_construction.ll:100]: At the moment the IR is a bit difficult to read. Could you rename the labels & value names to be more descriptive (and shorter/easier to read)? e.g. something like  loop.1.header, loop.1.latch, iv.1 and so on. (as  also mentioned in https://reviews.llvm.org/D150700)



<br class="Apple-interchange-newline" style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">

File: llvm/test/Transforms/LoopVectorize/outer_loop_hcfg_construction.ll [hunk 3]
```diff
@@ -0,0 +1,461 @@
...
+; CHECK-NEXT:    [[EXITCOND47_NOT:%.*]] = icmp eq i64 [[IV_NEXT43]], [[N]]
+; CHECK-NEXT:    br i1 [[EXITCOND47_NOT]], label [[FOR_COND_CLEANUP_LOOPEXIT:%.*]], label [[FOR_BODY_US]]
+; CHECK:       for.cond.cleanup.loopexit:
+; CHECK-NEXT:    br label [[FOR_COND_CLEANUP]]
+; CHECK:       for.cond.cleanup:
+; CHECK-NEXT:    ret void
+;
+entry:
+  %cmp34 = icmp sgt i64 %n, 0
+  br i1 %cmp34, label %for.body.us.preheader, label %for.cond.cleanup
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/outer_loop_hcfg_construction.ll:95]: should not be needed (as also mentioned in https://reviews.llvm.org/D150700)

File: llvm/test/Transforms/LoopVectorize/outer_loop_hcfg_construction.ll [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+
+for.body.us.preheader:
+  br label %for.body.us
+
+for.body.us:
+  %iv42 = phi i64 [ 0, %for.body.us.preheader ], [ %iv.next43, %for.cond2.for.cond.cleanup4_crit_edge.split.us.us ]
+  %arrayidx.us = getelementptr inbounds ptr, ptr %a, i64 %iv42
+  %0 = load ptr, ptr %arrayidx.us, align 8
+  store i32 0, ptr %0, align 4
+  %1 = trunc i64 %iv42 to i32
```
Comment (fhahn) [inline llvm/test/Transforms/LoopVectorize/outer_loop_hcfg_construction.ll:105]: should not be needed. (as also mentioned in https://reviews.llvm.org/D150700)

**Review by fhahn — COMMENTED** [review]
Thanks for moving this from Phabricator. Looks like there are a few outstanding comments from the original https://reviews.llvm.org/D150700.

**Comment by nikolaypanchenko** [comment]
> Thanks for moving this from Phabricator. Looks like there are a few outstanding comments from the original https://reviews.llvm.org/D150700.

my bad. Somehow missed our comments

File: llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp [hunk 5]
```diff
@@ -156,6 +156,19 @@ static bool isHeaderVPBB(VPBasicBlock *VPBB) {
   return VPBB->getParent() && VPBB->getParent()->getEntry() == VPBB;
 }
 
+/// Return true of \p L loop is contained within \p OuterLoop
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:159]: nit: period at end of comment.

File: llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp [hunk 6]
```diff
@@ -156,6 +156,19 @@ static bool isHeaderVPBB(VPBasicBlock *VPBB) {
...
+static bool isNestedLoop(const Loop *L, const Loop *OuterLoop) {
```
Comment (fhahn) [inline llvm/lib/Transforms/Vectorize/VPlanHCFGBuilder.cpp:160]: nit: might be better to call it something like `doesContainLoop`, more in line with the comment.

**Review by fhahn — APPROVED** [review]
LGTM thanks! 2 small remaining suggestions inline

**Review by fhahn — APPROVED** [review]
Still LGTM, thanks!

**Pull request merged**